vdr/store/bbolt.go

```diff
@@ -17,315 +17,321 @@ package store
+}
 
-		documents := tx.Bucket(documentsBucket)
-		if documents == nil {
-			return nil
+func (store *bboltStore) Write(document did.Document, metadata vdr.DocumentMetadata) error {
+	return store.db.Update(func(tx *bbolt.Tx) error {
+		if err := store.createBucketsIfNotExist(tx); err != nil {
+			return err
 		}
 
-		if err := versions.ForEach(func(key, data []byte) error {
-			versionList := parseDocumentVersionList(data)
-
-			doc, err := store.getDocumentVersion(documents, versionList.Latest())
-			if err != nil {
-				return err
-			}
+		latestBucket := tx.Bucket([]byte(latestBucket))
+		didString := document.ID.String()
 
-			if err := fn(doc.Document, doc.Metadata); err != nil {
-				return err
-			}
+		// first get latest
+		latestBytes := latestBucket.Get([]byte(didString))
+		if latestBytes != nil {
+			return vdr.ErrDIDAlreadyExists
+		}
 
-			return nil
-		}); err != nil {
-			return err
+		// add new metadata record pointing to latest
+		newMetadataRecord := metadataRecord{
+			Deactivated: IsDeactivated(document),
+			DID:         document.ID.String(),
+			Metadata:    metadata,
+			Version:     0,
 		}
 
-		return nil
+		return store.writeDocument(tx, document, newMetadataRecord)
 	})
 }
 
-func (store *bboltStore) filterDocument(doc *documentVersion, metadata *vdr.ResolveMetadata) error {
-	// Verify deactivated
-	if IsDeactivated(doc.Document) && (metadata == nil || !metadata.AllowDeactivated) {
-		return vdr.ErrDeactivated
-	}
-
-	if metadata == nil {
-		return nil
-	}
-
-	// Filter on hash
-	if metadata.Hash != nil && !doc.Metadata.Hash.Equals(*metadata.Hash) {
-		return vdr.ErrNotFound
-	}
+func (store *bboltStore) Update(id did.DID, current hash.SHA256Hash, next did.Document, metadata *vdr.DocumentMetadata) error {
+	return store.db.Update(func(tx *bbolt.Tx) error {
+		latestBucket := tx.Bucket([]byte(latestBucket))
+		metadataBucket := tx.Bucket([]byte(metadataBucket))
+		didString := id.String()
 
-	// Filter on creation and update time
-	if metadata.ResolveTime != nil {
-		resolveTime := *metadata.ResolveTime
+		// first get latest
+		var version int
+		var prevMetaRef []byte
+		latestRef := latestBucket.Get([]byte(didString))
 
-		if doc.Metadata.Created.After(resolveTime) {
+		// check for existence
+		if latestRef == nil {
 			return vdr.ErrNotFound
 		}
 
-		if doc.Metadata.Updated != nil {
-			if doc.Metadata.Updated.After(resolveTime) {
-				return vdr.ErrNotFound
-			}
+		latestBytes := metadataBucket.Get(latestRef)
+		latestMetadata := metadataRecord{}
+		if err := json.Unmarshal(latestBytes, &latestMetadata); err != nil {
+			return err
 		}
-	}
 
-	// Filter on SourceTransaction
-	if metadata.SourceTransaction != nil {
-		for i, keyTx := range doc.Metadata.SourceTransactions {
-			if keyTx.Equals(*metadata.SourceTransaction) {
-				break
-			}
-			if i == len(doc.Metadata.SourceTransactions)-1 {
-				return vdr.ErrNotFound
-			}
+		// check for deactivated
+		if latestMetadata.Deactivated {
```

Comment:
what if `latestMetadata` was a different branch in which it was deactivated, and this is an update, parallel to the TX that deactivated it? Theoretically the result is a non-conflicted document, and it an allowed update?

Reply:
Same goes for the hash. When reviewing current functionality, the checking of the hash is in the wrong place. It should be at the API level to prevent parallel updates. At the store level the source could be a TX from another node, checking the hash is irrelevant. It also exposes another issue: the latest DID Doc is retrieved in `ambassador.handleUpdateDIDDocument` while key fetching uses the prev transactions. That will have to be fixed first in another PR.

Reply:
Also added #879